server/src/main/java/io/crate/expression/scalar/NumTermsPerDocQuery.java

```diff
@@ -0,0 +1,237 @@
+import io.crate.types.BooleanType;
+import io.crate.types.ByteType;
+import io.crate.types.DataType;
+import io.crate.types.DoubleType;
+import io.crate.types.FloatType;
+import io.crate.types.GeoPointType;
+import io.crate.types.IntegerType;
+import io.crate.types.IpType;
+import io.crate.types.LongType;
+import io.crate.types.ShortType;
+import io.crate.types.StringType;
+import io.crate.types.TimestampType;
+
+public class NumTermsPerDocQuery extends Query {
+
+    private final String column;
+    private final java.util.function.Function<LeafReaderContext, IntUnaryOperator> numTermsPerDocFactory;
+    private final IntPredicate matches;
+
+    public static NumTermsPerDocQuery forArray(Reference arrayRef, IntPredicate valueCountIsMatch) {
+        return new NumTermsPerDocQuery(
+            arrayRef.column().fqn(),
+            leafReaderContext -> getNumTermsPerDocFunction(leafReaderContext.reader(), arrayRef),
+            valueCountIsMatch
+        );
+    }
+
+    private static IntUnaryOperator getNumTermsPerDocFunction(LeafReader reader, Reference ref) {
+        DataType<?> elementType = ArrayType.unnest(ref.valueType());
+        switch (elementType.id()) {
+            case BooleanType.ID:
+            case ByteType.ID:
+            case ShortType.ID:
+            case IntegerType.ID:
+            case LongType.ID:
+            case TimestampType.ID_WITH_TZ:
+            case TimestampType.ID_WITHOUT_TZ:
+            case FloatType.ID:
+            case DoubleType.ID:
+            case GeoPointType.ID:
+                return numValuesPerDocForSortedNumeric(reader, ref.column());
+
+            case StringType.ID:
+                return numValuesPerDocForString(reader, ref.column());
+
+            case IpType.ID:
+                return numValuesPerDocForIP(reader, ref.column());
+
+            default:
+                throw new UnsupportedOperationException("NYI: " + elementType);
+        }
+    }
+
+    private static IntUnaryOperator numValuesPerDocForIP(LeafReader reader, ColumnIdent column) {
+        SortedSetDocValues docValues;
+        try {
+            docValues = reader.getSortedSetDocValues(column.fqn());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return doc -> {
+            try {
+                if (docValues.advanceExact(doc)) {
+                    int c = 1;
+                    while (docValues.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                        c++;
+                    }
+                    return c;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return 0;
+        };
+    }
+
+    private static IntUnaryOperator numValuesPerDocForString(LeafReader reader, ColumnIdent column) {
+        SortedBinaryDocValues docValues;
+        try {
+            docValues = FieldData.toString(DocValues.getSortedSet(reader, column.fqn()));
```

Comment:
Also unrelated to this PR but it might be possible to remove the FieldData.toString call,  and count the terms without it.